File: python/68_magic_5gon_ring.py
from itertools import permutations # built-in library for generating permutations of iterable with unique values

def magic_5gon_ring() -> str:
    
    max_str = "0"
    for ring in permutations(range(1, 11)):
        if not is_valid_ring(ring):
            continue
        digit_str = get_digit_string(ring)
        max_str = digit_str if int(digit_str) > int(max_str) else max_str
            
    return str(max_str)
    
def is_valid_ring(ring: list) -> bool:
    inner = ring[:5]
    outer = ring[5:]
    
    # valid strings start from the group of three with the numerically lowest external node
    if outer[0] != min(outer):
        return False
    
    # digit string must have 16 digits
    if len(get_digit_string(ring)) != 16:
        return False
        
    # each group of three must have the same sum
    if len(set(outer[i] + inner[i] + inner[(i + 1) % 5] for i in range(0, 5))) != 1:
        return False

    return True
    
def get_digit_string(ring: list) -> str:
    inner = ring[:5]
    outer = ring[5:]
    
    # [a, b, c, d, e, f, g, h, i, j]
    # inner: [a, b, c, d, e]
    # outer: [f, g, h, i, j]
    # string: fabgbchcdidejea
    
    digit_str = ""
    for i in range(0, 5):
        digit_str += str(outer[i]) + str(inner[i]) + str(inner[(i + 1) % 5])
    return digit_str
    
if __name__ == "__main__":
    print(magic_5gon_ring())

File: cpp/68_magic_5gon_ring.cpp
#include <iostream>
#include <vector>
#include <algorithm>

bool check_sums(std::vector<int> outer, std::vector<int> inner){
    int sum = outer[4] + inner[4] + inner[0];
    for (int i = 0; i < 4; i++){
        if (outer[i] + inner[i] + inner[i + 1] != sum) return false;
    }
    return true;
}

std::string stringify(std::vector<int> outer, std::vector<int> inner){
    std::string res = "";
    for (int i = 0; i < 4; i++){
        res += std::to_string(outer[i]) + std::to_string(inner[i]) + std::to_string(inner[i + 1]);
    }
    res += std::to_string(outer[4]) + std::to_string(inner[4]) + std::to_string(inner[0]);
    return res;
}

std::string magic_5gon_ring(){
    std::string res = "";
    // for string representation to have 16 digits, 10 must be an external node
    std::vector<int> outer = {6, 7, 8, 9, 10};
    // only consider string representations starting with numerically lowest external node
    // permute outer ring until outer[0] becomes 7
    while (outer[0] == 6){
        std::vector<int> inner = {1, 2, 3, 4, 5};
        // consider each permutation of inner ring
        while(std::next_permutation(inner.begin(), inner.end())){
            // check that each line of nodes has the same sum
            if (check_sums(outer, inner)){
                // all valid string representations are 16 digits
                // lexicographically last and numerically greatest are equivalent here
                res = max(res, stringify(outer, inner));
            }
        }
        std::next_permutation(outer.begin(), outer.end());
    }
    return res;
}

int main(){
    std::cout << magic_5gon_ring() << "\n";
}
